tavern/cli/tomeuploader/main.go

```diff
@@ -0,0 +1,308 @@
+	if basedir == "" {
+		basedir = "tomes"
+	}
+	endpoint := os.Getenv("endpoint")
+	graphqlEndpoint := ""
+	if endpoint == "" {
+		graphqlEndpoint = "http://localhost/graphql"
+		endpoint = "http://127.0.0.1"
+	} else {
+		graphqlEndpoint = endpoint + "/graphql"
+	}
+	cookie := os.Getenv("cookie")
+	if cookie == "" {
+		ctx, cancel := context.WithTimeout(context.Background(), 2*time.Minute)
+		defer cancel()
+
+		// Configure Browser (uses the default system browser)
+		browser := auth.BrowserFunc(browser.OpenURL)
+
+		// Open Browser and Obtain Access Token (via 127.0.0.1 redirect)
+		token, err := auth.Authenticate(ctx, browser, endpoint)
+		if err != nil {
+			panic(err)
+		}
+		cookie = token.String()
+	}
+
+	// Call the function to upload tomes
+	if err := UploadTomesGraphQL(basedir, graphqlEndpoint, cookie); err != nil {
+		fmt.Println("Error:", err)
+	}
+}
+
+// UploadTomesGraphQL uploads tomes using GraphQL.
+func UploadTomesGraphQL(basedir string, graphqlEndpoint string, cookie string) error {
+	dir, _ := os.ReadDir(basedir)
+	entries, err := os.ReadDir(basedir)
+	if err != nil {
+		return fmt.Errorf("failed to read filesystem: %w", err)
+	}
+
+	for _, entry := range entries {
+		if !entry.IsDir() {
+			continue
+		}
+
+		tomeName := entry.Name()
+
+		// Check if the tome already exists
+		if exists, err := checkTomeExists(graphqlEndpoint, tomeName, cookie); err != nil {
+			return err
+		} else if exists {
+			//todo add updating the existing tome
+			continue
+		}
+
+		// ... inside the loop over entries
+		metadata, eldritchContent, tomeFiles, err := processTomeDirectory(dir, basedir+"/"+tomeName)
+		if err != nil {
+			return err
+		}
+
+		// Create the tome using GraphQL with the metadata and eldritchContent
+		if err := createTomeGraphQL(graphqlEndpoint, metadata, eldritchContent, tomeFiles, cookie); err != nil {
+			return err
+		}
+		// ...
+
+		// // Create the tome using GraphQL
+		// if err := createTomeGraphQL(graphqlEndpoint, metadata, tomeName, tomeFiles); err != nil {
+		// 	return err
+		// }
+	}
+
+	return nil
+}
+
+// checkTomeExists checks if a tome already exists using a GraphQL query.
+func checkTomeExists(endpoint, tomeName string, cookie string) (bool, error) {
+	query := fmt.Sprintf(`query CheckTomeExists {
+		tomes(where: {
+			name: "%s"
+	
+		}) {
+			id
+		}
+	}
+	`, tomeName)
+	response, err := sendGraphQLRequest(endpoint, query, cookie)
+	if err != nil {
+		return false, err
+	}
+
+	// Extract the 'exists' field from the response
+	var result struct {
+		Data struct {
+			Tomes []string `json:"tomes"`
+		}
+	}
+	if err := json.Unmarshal(response, &result); err != nil {
+		return false, fmt.Errorf("failed to parse GraphQL response: %w", err)
+	}
+	if len(result.Data.Tomes) > 0 {
+		return true, nil
+	}
+	return false, nil
+}
+
+// processTomeDirectory reads files in a tome directory and prepares them for upload.
+func processTomeDirectory(fs []fs.DirEntry, dirName string) (Metadata, string, []File, error) {
```

Comment:
We can likely reuse some functionality from the "realm.pub/tavern/tomes" package here